app/assets/javascripts/components/uploads/uploads_handler.jsx

```diff
@@ -13,10 +14,37 @@ const UploadsHandler = createReactClass({
     course: PropTypes.object
   },
 
+  getInitialState() {
+    return {
+      offset: 0,
+      data: [],
+      perPage: 100,
+    };
+  },
+
   componentWillMount() {
     return this.props.receiveUploads(this.props.course_id);
   },
 
+  componentWillReceiveProps(nextProps) {
+    const data = nextProps.uploads.slice(this.state.offset, this.state.offset + this.state.perPage);
+    this.setState({ data: data });
+    this.setState({ pageCount: nextProps.uploads.length / this.state.perPage });
+  },
+
+  setUploadData() {
+    const data = this.props.uploads.slice(this.state.offset, this.state.offset + this.state.perPage);
+    this.setState({ data: data });
+  },
+
+  handlePageClick(data) {
+    const selectedPage = data.selected;
+    const offset = Math.ceil(selectedPage * this.state.perPage);
+    this.setState({ offset: offset }, () => {
+      this.setUploadData();
```

Comment:
Is there a specific reason you are using a callback to update the state for new `data`?

Reply:
Yes, or else there's an unwanted effect, the `setUpload` executes before the state is actually set

Reply:
Well actually, I thought you could set the state with only one call to `setState` with `offset` and `data`. Just compute the `data` array after you calculate the `offset`  and then pass them together to `setState`. I just thought right now, this will rerender the component by setting the state again in the callback.

Reply:
Surely, I can pass offset to `setUploadData` and then set the state together after computing the data.